### ISSUE TITLE
Severe SDPA Performance Regression 2.5.0-RC1

### ISSUE THREAD

**Issue by Beinsezii** [pytorch/pytorch#135778]
### 🐛 Describe the bug

Performance on my 7900 XTX is almost cut in half for any kind of inference (ex: [diffusers](https://github.com/huggingface/diffusers)) on what appears to be `torch.nn.functional.scaled_dot_product_attention`. I assume the cause is https://github.com/pytorch/pytorch/pull/128922 which unconditionally force-upcasts everything to `torch.float32`. The memory usage reflects this, up to 12.0GiB from 9.5GiB on 2.4.1 for a 4x128x128 latent shape.

Another AMD user has experienced the same https://github.com/ROCm/aotriton/pull/39#issuecomment-2333063812

~~It probably also affects MPS https://github.com/pytorch/pytorch/pull/128922#issuecomment-2344823135~~
MPS has a different sdpa path and is likely a [different bug](https://github.com/pytorch/pytorch/issues/136003)

The impact may be as wide as all other platforms that have fp16 intrinsics but no FLASH_ATTENTION or EFFICIENT_ATTENTION sdpa kernels which is effectively every non-NVIDIA gpu from the last few years. I do not have access to other hardware types that may confirm this.

Rough example:
```python
import torch
import time

torch.set_default_device("cuda")
torch.set_default_dtype(torch.float16)

SHAPE = (8, 64, 1024, 1024)

@torch.inference_mode()
def benchmark(n: int) -> float:
    now = time.perf_counter()
    for _ in range(n):
        torch.nn.functional.scaled_dot_product_attention(torch.randn(SHAPE), torch.randn(SHAPE), torch.randn(SHAPE))
    return time.perf_counter() - now

benchmark(100)  # warmup
print(f"100 steps of SDPA completed in {benchmark(100) * 1000}ms")
```
On Torch 2.4.1+rocm6.1 it completes 100 steps in ≈3.67ms, while 2.5.0-rc1+rocm6.1 takes ≈7.16ms

A more involved benchmark at https://github.com/ROCm/aotriton/pull/39#issuecomment-2330669978 also shows significant MATH losses for 2.5.0.

### Versions
I've confirmed the regression on both Ubuntu docker using the official ROCm packages and on Arch Linux natively.

`python -c 'import torch; torch.utils.collect_env.main()'` output:
```
Collecting environment information...
PyTorch version: 2.5.0+rocm6.1
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.1.40091-a8dbc0c19

OS: Ubuntu 24.04 LTS (x86_64)
GCC version: (Ubuntu 13.2.0-23ubuntu4) 13.2.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jul 31 2024, 17:43:48) [GCC 13.2.0] (64-bit runtime)
Python platform: Linux-6.10.9-arch1-2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Radeon RX 7900 XTX (gfx1100)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.1.40091
MIOpen runtime version: 3.1.0
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               24
On-line CPU(s) list:                  0-23
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7900X 12-Core Processor
CPU family:                           25
Model:                                97
Thread(s) per core:                   2
Core(s) per socket:                   12
Socket(s):                            1
Stepping:                             2
CPU(s) scaling MHz:                   49%
CPU max MHz:                          5908.0000
CPU min MHz:                          400.0000
BogoMIPS:                             9404.15
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d amd_lbr_pmc_freeze
Virtualization:                       AMD-V
L1d cache:                            384 KiB (12 instances)
L1i cache:                            384 KiB (12 instances)
L2 cache:                             12 MiB (12 instances)
L3 cache:                             64 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-23
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] pytorch-triton-rocm==3.0.0+757b6a61e7
[pip3] torch==2.5.0+rocm6.1
[pip3] torchao==0.4.0
[pip3] triton==3.0.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @drisspg @mikaylagawarecki

**Comment by xw285cornell** [comment]
This was quite extensively debated and it's essentially a trade off between accuracy vs. speed, especially if you run the same code across different HW. In Nvidia (or even AMD GPU) today, there is MEA or FA implementation; and others are using MATH backend. There will be a numerical difference (of course depending on the input range) due to the fusion/non-fusion of the ops within SDPA, which took us a lot of time to debug what went on. 
One possibility is to do pre-op casting, however this makes the user code sort of ugly, where you'll need to check something like "can_use_flash_attention" for the casting op. 
With all that said, I agree that force fp32 w/o an option to convert back maybe too brutal. So i wonder if we can make this configurable.

**Comment by Beinsezii** [comment]
> This was quite extensively debated and it's essentially a trade off between accuracy vs. speed, especially if you run the same code across different HW. In Nvidia (or even AMD GPU) today, there is MEA or FA implementation

AMD does not have a suitable FA implementation to justify the changes to MATH. FA is partially supported on the two newest GFX9 enterprise cards and with https://github.com/pytorch/pytorch/pull/135869 it will be very experimentally supported on GFX1100 with less speed than 2.4's MATH. You can see @evshiron's extensive benchmarks in the AOTriton thread.
Not to mention GFX1101, GFX10, probably some older GFX9 cards that also have hardware fp16. Unless FlexAttention or some other replacement matures enough replace MATH globally I don't see how full precision is ever a justifiable tradeoff.

If the expectation from the NVIDIA PoV is to manually switch from FA to MATH for high precision ops, perhaps MATH should be split into two separate kernels, one for speed and one for precision.

**Comment by jianyuh** [comment]
> AMD does not have a suitable FA implementation to justify the changes to MATH. 

BTW, AMD has a FA implementation in xformers ( https://github.com/facebookresearch/xformers/tree/main/xformers/csrc/attention/hip_fmha ), based on Composable Kernel, working for MI300x / MI250. It might take some efforts to port it back to PyTorch SDPA.

**Comment by evshiron** [comment]
> BTW, AMD has a FA implementation in xformers ( https://github.com/facebookresearch/xformers/tree/main/xformers/csrc/attention/hip_fmha ), based on Composable Kernel, working for MI300x / MI250. It might take some efforts to port it back to PyTorch SDPA.

CDNA devices are excellent and they deserve a lot of support. In fact, [Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention) can already work on MI250/MI300 with good performance.

However, to support more devices easily, AMD is using AOTriton for PyTorch integration, which archives kernels generated by Triton based on presets.

This approach highly depends on Triton compiler and LLVM optimizations, and doesn't really match the performance of CK-based ones. For example, it's about 30% slower than PyTorch's Math SDPA when doing Stable DIffusion on my RX 7900 XTX. That's why we hope the performance of the Math implementation remains.

**Comment by hvaara** [comment]
MPS has its own math implementation in https://github.com/pytorch/pytorch/blob/2f53d570febc6afe69473f0e4989ff59342edb57/aten/src/ATen/native/mps/operations/Attention.mm, so AFAICT it's not impacted by the changes in #128922.

**Comment by kit1980** [comment]
Looks like the fix is not ready?
@janeyx99 do you think 2.5 milestone is still relevant?

**Comment by kit1980** [comment]
This should be fixed by https://github.com/pytorch/pytorch/pull/136536
